### PR TITLE
Implement "event_fields" in filters

### PR DESCRIPTION
**Pull request by Kegsay**
http://matrix.org/docs/spec/client_server/r0.2.0.html#post-matrix-client-r0-user-userid-filter

```
event_fields	[string]
```
> List of event fields to include. If this list is absent then all fields are included. The entries may include '.' charaters to indicate sub-fields. So ['content.body'] will include the 'body' field of the 'content' object. A literal '.' character in a field name may be escaped using a '\'. A server may include more fields than were requested.

### CODE REVIEW

File: synapse/events/utils.py [hunk 1]
```diff
@@ -97,6 +108,87 @@ def add_fields(*fields):
...
+        return dictionary
+
+    # for each field, convert it:
+    # ["content.body.thing\.with\.dots"] => [["content", "body", "thing\.with\.dots"]]
+    split_fields = [SPLIT_FIELD_REGEX.split(f) for f in fields]
+
+    # for each element of the output array of arrays:
+    # remove escaping so we can use the right key names. This purposefully avoids
+    # using list comprehensions to avoid needless allocations as this may be called
+    # on a lot of events.
```
Comment (erikjohnston) [inline synapse/events/utils.py]: I did a quick benchmark and the following list comprehension is actually quicker:

```python
split_fields[:] = [ [field.replace(r'\.', r'.') for field in field_array] for field_array in split_fields]
```

File: synapse/events/utils.py [hunk 2]
```diff
@@ -164,6 +256,10 @@ def serialize_event(e, time_now_ms, as_client_event=True,
                 d["unsigned"]["transaction_id"] = txn_id
 
     if as_client_event:
-        return event_format(d)
-    else:
-        return d
+        d = event_format(d)
+
+    if (only_event_fields and isinstance(only_event_fields, list) and
+            all(isinstance(f, basestring) for f in only_event_fields)):
```
Comment (erikjohnston) [inline synapse/events/utils.py]: If the args are incorrect this should raise an exception.

File: synapse/events/utils.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/events/utils.py]: And if we really cared about performance, we would pull this up to be calculated when we initially parsed the filter

File: synapse/events/utils.py (hunk 2 above)
Comment (Kegsay) [inline synapse/events/utils.py]: Done.

File: synapse/events/utils.py (hunk 1 above)
Comment (Kegsay) [inline synapse/events/utils.py]: Done.

File: synapse/events/utils.py [hunk 3]
```diff
@@ -97,6 +108,87 @@ def add_fields(*fields):
...
+        return
+
+    # Insert the key into the output dictionary, creating nested objects
+    # as required. We couldn't do this any earlier or else we'd need to delete
+    # the empty objects if the key didn't exist.
+    sub_out_dict = dst
+    for sub_field in field:
+        if sub_field not in sub_out_dict:
+            sub_out_dict[sub_field] = {}
+        sub_out_dict = sub_out_dict[sub_field]
```
Comment (Kegsay) [inline synapse/events/utils.py]: Done.

**Comment by erikjohnston** [comment]
LGTM

**Pull request merged**